### PR TITLE
New Docker build configuration

### PR DESCRIPTION
**Pull request by schustmi**
## Describe changes

This PR greatly increases the flexibility in which users can specify how ZenML builds the Docker image that is used to run their pipeline steps in remote orchestrators/step operators. Additionally, it reduces the code duplication of the Docker image building code that was found in many stack components by introducing a simple pydantic mixin `PipelineDockerImageBuilder` which all stack components can inherit from to easily build Docker images.

## Pre-requisites
Please ensure you have done the following:
- [x] I have read the **CONTRIBUTING.md** document.
- [x] If my change requires a change to docs, I have updated the documentation accordingly.
- [ ] If I have added an integration, I have updated the [integrations](https://docs.zenml.io/advanced-guide/integrations) table and the [corresponding website section](https://zenml.io/integrations).
- [x] I have added tests to cover my changes.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Other (add details above)

### CODE REVIEW

File: docs/book/developer-guide/advanced-usage/docker.md [hunk 1]
```diff
@@ -31,60 +41,155 @@ following two ways:
...
+
+{% hint style="warning" %}
+This is an advanced feature and will most likely break your pipelines. If you use this,
+you're on your own and need to copy a stack configuration to the correct path yourself.
+{% endhint %}
+
+### How to install additional pip dependencies
+
+By default, ZenML will automatically install all the packages required by your
+active ZenML stack. There are however many additional ways in which you can specify
```
Comment (bcdurak) [inline docs/book/developer-guide/advanced-usage/docker.md:88]:
```suggestion
active ZenML stack. There are, however, various ways in which you can specify
```

File: src/zenml/pipelines/base_pipeline.py [hunk 2]
```diff
@@ -136,23 +134,50 @@ def __init__(self, *args: BaseStep, **kwargs: Any) -> None:
...
+            kwargs.pop(PARAM_DOCKER_CONFIGURATION, None)
+            or DockerConfiguration()
+        )
+        self._migrate_to_docker_config(kwargs)
+
         self.name = self.__class__.__name__
         self.__steps: Dict[str, BaseStep] = {}
         self._verify_arguments(*args, **kwargs)
 
+    def _migrate_to_docker_config(self, kwargs: Dict[str, Any]) -> None:
```
Comment (bcdurak) [inline src/zenml/pipelines/base_pipeline.py:149]: The docstring for the method is missing here.

File: src/zenml/steps/step_environment.py [hunk 3]
```diff
@@ -102,10 +105,10 @@ def cache_enabled(self) -> bool:
         return self._cache_enabled
 
     @property
-    def pipeline_requirements(self) -> List[str]:
-        """The name of the currently running step.
+    def docker_configuration(self) -> "DockerConfiguration":
```
Comment (bcdurak) [inline src/zenml/steps/step_environment.py:108]: Is this used anywhere? I am not sure if I am missing anything.

File: src/zenml/orchestrators/base_orchestrator.py [hunk 4]
```diff
@@ -505,10 +505,12 @@ def _configure_node_context(
             runtime_config_copy, sort_keys=True, default=pydantic_encoder
         )
 
+        docker_config_json = pipeline.docker_configuration.json(sort_keys=True)
```
Comment (bcdurak) [inline src/zenml/orchestrators/base_orchestrator.py:508]: Just out of curiosity, I see two different paradigms here. While the `runtime_configuration` is given to this function by the orchestrator, the `docker_configuration` is treated as a class variable of the pipeline. Is there a specific reason why it is like this?

**Review by bcdurak — COMMENTED** [review]
I have been reading this for a while and everything looks amazing. I have left a few small comments.

I am so looking forward to merging this into my branch. I will try it right away and share my findings from a more hands-on review process :)

File: src/zenml/pipelines/base_pipeline.py (hunk 2 above)
Comment (schustmi) [reply src/zenml/pipelines/base_pipeline.py:149]: Good catch, I added it!

File: src/zenml/steps/step_environment.py (hunk 3 above)
Comment (schustmi) [reply src/zenml/steps/step_environment.py:108]: Not in my PR, but @safoinme added the `pipeline_requirements` property to the step environment so he could then use it in this custom code deployment PR. Now the pipeline requirements don't exist anymore but got replaced by the docker config.

File: src/zenml/orchestrators/base_orchestrator.py (hunk 4 above)
Comment (schustmi) [reply src/zenml/orchestrators/base_orchestrator.py:508]: Mhm correct, I think this is mainly due to the `docker_configuration` being specified on the `@pipeline` decorator right now, while the runtime config is passed to `.run(...)`. However I think the entire configurations are quite messy at the moment, hopefully the design document we're creating this week will bring some more clarity in how we handle these things in the future :)

**Review by safoinme — COMMENTED** [review]
Half through The PR, such an amazing implementation to something that really bothered me while working on the custom deployment!
@schustmi one question regarding the additional build argument, what would be the best way to add an additional argument (e.g. `expose 5000`)

**Comment by schustmi** [comment]
@safoinme I'm assuming that by build argument in this case you mean adding an additional line to the generated Dockfile. 

Here would be one potential solution how to do that:

```python
MyModelDeployer(BaseModelDeployer, PipelineDockerImageBuilder):
    @staticmethod
    def _generate_zenml_pipeline_dockerfile(
        parent_image: str,
        docker_configuration: DockerConfiguration,
        requirements_files: Sequence[str] = (),
        entrypoint: Optional[str] = None,
    ) -> List[str]:
        lines  = super()._generate_zenml_pipeline_dockerfile(parent_image, docker_configuration, requirements_files, entrypoint)
        lines.append("EXPOSE 5000")
        return lines
```

File: src/zenml/orchestrators/base_orchestrator.py (hunk 4 above)
Comment (bcdurak) [reply src/zenml/orchestrators/base_orchestrator.py:508]: Looking forward to it :)

**Review by safoinme — APPROVED** [review]
LGTM!! Looking forward to see this merged

**Pull request merged**